### ISSUE TITLE
[Feature Request]: newtypes for arbitrary inner types

### ISSUE THREAD

**Issue by greyblake** [greyblake/nutype#7]
## Currently 

Currently `#[nutype]` macro can be applied only on predefined set of types: integers, floats, `String`.

It enables defining specific sanitizers/validators depending on the inner type.

## Feature request

However, there is some demand to use `#[nutype]` for other (unknown to nutype) types.
It must be still possible to define custom sanitizers and validators with `with = `.

## Related issues

* https://github.com/greyblake/nutype/issues/4

**Comment by Proksima** [comment]
First off, amazing crate, really making working with new types a breeze. Thank you!

Thought I would share one use case where I thought I would use this crate before I realised arbitrary types were not supported: to maintain invariants on some polars' DataFrames.

By defining custom validators which validate the shape, column names, presence of None/NaN, etc. Would allow to pass wrapped DataFrame around with confidence.

This is particularly helpful as transformation on data frames usually involve multiple chained operations (using the lazy-API) which makes a computational mistake "corrupting" the DataFrame a frequent issue when working with polars. This crate has the potential to offer a nice QoL improvement for this workflow I believe.

Anyway, thanks again for the great crate!

**Comment by greyblake** [comment]
@Proksima Hi!
Thanks for your feedback! I have a plan to add support for arbitrary types, however there will be a couple of limitations:
* There are will be no predefined sanitizers or validators (only custom ones)
* Because of that it won't be possible to derive `arbitrary::Arbitrary` for those types

**Comment by fenollp** [comment]
What do you think about supporting `Vec`?
```rust
#[nutype(validate(sorted))]
#[derive(Clone, Ord)]
struct MySortedUsizeVec(Vec<usize>);
```
where `sorted` could be something akin to
```rust
use itertools::Itertools;

fn is_sorted<I>(data: I) -> bool
where
    I: IntoIterator,
    I::Item: Ord + Clone,
{
    data.into_iter().tuple_windows().all(|(a, b)| a <= b)
}
```
I'm wondering how to express `mut self Vec` operations then. Should `nutype` make all of `Vec` methods directly available from the outer type? Should validation happen right after each of these calls? What about introducing methods that behave a bit differently based on the validation properties? `v.push(42)` would then place `42` at the right spot in the inner vec.

**Comment by greyblake** [comment]
@fenollp  I have a further plan to have some special support of collections (not only `Vec`).

However, the support of arbitrary types will be first. Having that in place users will be able to wrap collections and implement their custom validations.

Speaking of mutability (`v.push(42)`) it maybe be bit problematic, cause we need to be sure that constraints are not violated. We probably can have mutable operations, but they will be fallible.

**Comment by greyblake** [comment]
Addressed in https://github.com/greyblake/nutype/pull/88

**Comment by greyblake** [comment]
@fenollp 

So, here is what you can achieve with Vec at the moment:

```rs
use nutype::nutype;

#[nutype(
    derive(Debug, PartialEq, Deref, AsRef),
    sanitize(with = |mut guests: Vec<String>| { guests.sort(); guests }),
    validate(predicate = |guests| !guests.is_empty() ),
)]
pub struct GuestList(Vec<String>);

fn main() {
    // Empty list is not allowed
    assert_eq!(
        GuestList::new(vec![]),
        Err(GuestListError::PredicateViolated)
    );

    // Create the list of our guests
    let guest_list = GuestList::new(vec![
        "Seneca".to_string(),
        "Marcus Aurelius".to_string(),
        "Socrates".to_string(),
        "Epictetus".to_string(),
    ]).unwrap();

    // The list is sorted (thanks to sanitize)
    assert_eq!(
        guest_list.as_ref(),
        &[
            "Epictetus".to_string(),
            "Marcus Aurelius".to_string(),
            "Seneca".to_string(),
            "Socrates".to_string(),
        ]
    );

    // Since GuestList derives Deref, we can use methods from `Vec<T>`
    // due to deref coercion (if it's a good idea or not, it's left up to you to decide!).
    assert_eq!(guest_list.len(), 4);

    for guest in guest_list.iter() {
        println!("{guest}");
    }
}

```


Note, you still cannot `push`, because the GuestList is immutable.

**Comment by greyblake** [comment]
The functionality is available in `0.4.0-beta.1` at the moment:
https://crates.io/crates/nutype/0.4.0-beta.1
